test/integration/forms/patient.js

```diff
@@ -0,0 +1,806 @@
+      },
+    }));
+
+    cy
+      .routeForm(_.identity, '11111')
+      .routeFormDefinition()
+      .routePatient(fx => {
+        fx.data.id = '1';
+        return fx;
+      })
+      .routeFormFields(fx => {
+        fx.data.attributes = { patient: { fields: { foo: 'bar' } } };
+
+        return fx;
+      })
+      .visit('/patient/1/form/11111')
+      .wait('@routeForm')
+      .wait('@routePatient');
+
+    cy
+      .get('.form__controls')
+      .find('.form__last-updated')
+      .should('contain', 'Your work is stored automatically.')
+      .should('contain', `Last edit was ${ formatDate(testTs(), 'AGO_OR_TODAY') }`);
+
+    cy
+      .get('.form__content')
+      .should('contain', `Last edit was ${ formatDate(testTs(), 'TIME_OR_DAY') }`)
+      .find('.js-discard')
+      .click();
+
+    cy
+      .get('.modal--small')
+      .find('.js-submit')
+      .click();
+
+    cy
+      .wait('@routeFormDefinition')
+      .wait('@routeFormFields');
+
+    cy
+      .get('.form__controls')
+      .find('.form__last-updated')
+      .should('contain', 'Your work is stored automatically.')
+      .should('not.contain', `Last edit was ${ formatDate(testTs(), 'AGO_OR_TODAY') }`);
+
+    cy
+      .iframe()
+      .find('[name="data[patient.fields.foo]"]')
+      .should('have.value', 'bar');
+  });
+
+
+  specify('read only form', function() {
+    localStorage.setItem('form-subm-11111-1-22222', JSON.stringify({
+      updated: testTs(),
+      submission: {
+        patient: { fields: { foo: 'foo' } },
+      },
+    }));
+
+    cy
+      .routePatient(fx => {
+        fx.data.id = '1';
+        return fx;
+      })
+      .routeForm(_.identity, '22222')
+      .routeFormDefinition()
+      .routeFormFields(fx => {
+        fx.data.attributes = { patient: { fields: { foo: 'bar' } } };
+
+        return fx;
+      })
+      .visit('/patient/1/form/22222')
+      .wait('@routeForm')
+      .wait('@routePatient')
+      .wait('@routeFormDefinition')
+      .wait('@routeFormFields');
+
+    cy
+      .get('[data-form-updated-region]')
+      .should('be.empty');
+
+
+    cy
+      .get('.form__controls')
+      .find('button')
+      .contains('Read Only')
+      .should('be.disabled');
+
+    cy
+      .get('.form-widgets')
+      .should('not.exist');
+
+    cy
+      .iframe()
+      .find('[name="data[patient.fields.foo]"]')
+      .should('have.value', 'bar');
+  });
+
+  specify('form scripts and reducers', function() {
+    cy
+      .routePatient(fx => {
+        fx.data.id = '1';
+        return fx;
+      })
+      .routeForm(_.identity, '33333')
+      .routeFormDefinition()
+      .routeFormFields()
+      .visit('/patient/1/form/33333')
+      .wait('@routePatient')
+      .wait('@routeForm')
+      .wait('@routeFormFields')
+      .wait('@routeFormDefinition');
+
+    cy
+      .iframe()
+      .find('textarea[name="data[storyTime]"]')
+      .should('contain', 'foo');
+  });
+
+  if (Cypress.browser.name !== 'firefox') {
```

Comment:
Do we still need this? Can't remember why added it in the first place...

Reply:
I don't think you can stub `console.foo` in firefox